src/core/net/dns_headers.hpp

```diff
@@ -374,6 +374,182 @@ class Header : public Clearable<Header>
 
 } OT_TOOL_PACKED_END;
 
+/**
+ * This class implement helper methods for encoding/decoding of DNS Domain Names.
+ *
+ */
+class DomainName
+{
+public:
+    enum : uint8_t
+    {
+        kMaxLabelLength = 63,  ///< Max number of characters in a label.
+        kMaxLength      = 255, ///< Max number of characters in a domain name.
+    };
+
+    /**
+     * This static method encodes and appends a single domain name label to a message.
+     *
+     * The @p aLabel is assumed to contain a single domain name label as a C string (null-terminated). Unlike
+     * `AppendMultipleLabels()` which parses the label string and treats it as sequence of multiple (dot-separated)
+     * labels, this method always appends @p aLabel as a single whole label. This allows the label string to even
+     * contain dot '.' character, which, for example, is useful for "Service Instance Names" where <Instance> portion
+     * is a user-friendly name and can contain dot character.
+     *
+     * @param[in] aLabel              The label string to append. MUST not be nullptr.
+     * @param[in] aMessage            The message to append to.
```

Comment:
```suggestion
     * @param[out] aMessage            The message to append to.
```

Reply:
Or maybe we should call it `inout`? And other occurrences of `aMessage` in API docs.

Reply:
We seem to use `@param[in]` in a lot of other similar situations where things in a `aMessage` (or in general an object) can change by the method/function.  Examples in the code:
- `Mle::AppendHeader(Message &aMessage)`, `AppendMode()`, and a bunch more in `Mle`,
- In `Tlv` class we have a lot of `Tlv::Append(Message &aMessage)` flavors,
- `Dataset::AppendMleDatasetTlv(Message &aMessage)`,
- `MessageQueue::Enqueue()/Dequeue()` also refer to `aMessage` as `@param[in]` (though they technically change some state in the `aMessage` object).

For consistency, I'd prefer to keep it the same in this PR.

----

We can discuss whether it makes sense to change all other occurrences (from a different PR probably):

Here is my preference on this:
- I would say `@param[out]` would be good when variable/object is changed fully (e.g., a new instance of object is created/used or object is modified completely).
- For `Append()` like of methods I would prefer to continue to use `@param[in]` over `[inout]` (if we go for `[inout]` the situations for methods like `Enqueue()/Dequeue()` is harder to consider).

Thoughst? @jwhui, others.

Reply:
SGTM.